File: scripts/lint.js
// scripts/lint.js

const fs = require("fs");
const path = require("path");

const ROOT = process.cwd();

// Config
const PROTECTED_PATHS = [
  "_config.yml",
  "_data",
  "_layouts",
  "scripts",
  "README.md",
  "CODEOWNERS"
];

const IGNORE_DIRS = [
  ".git",
  ".github",
  "_data",
  "_layouts",
  "scripts",
  "node_modules"
];

const IGNORE_FILES = [
  "_config.yml",
  "README.md",
  "CODEOWNERS",
  "changed_files.txt"
];

const FILE_NAME_REGEX = /^\d{3}-[a-z0-9-]+$/;

let errors = [];
let warnings = [];

// Helpers
function isMarkdown(file) {
  return file.endsWith(".md");
}

function isEnglishName(name) {
  return /^[a-z0-9\-_.]+$/i.test(name);
}

function checkFileName(filePath) {
  const name = path.basename(filePath);

  // Allow index.md everywhere
  if (name === "index.md") return;

  const base = path.basename(filePath, ".md");

  // Allow:
  // 010-xxx
  // 01.xxx (optional if you want)
  const validPattern1 = /^\d{3}-[a-z0-9-]+$/;
  const validPattern2 = /^\d+\.[a-z0-9-]+$/;

  if (!validPattern1.test(base) && !validPattern2.test(base)) {
    errors.push(`❌ Invalid file name: ${filePath}`);
  }
}

function shouldCheckIndex(dirPath) {
  return !IGNORE_DIRS.some(d => dirPath.includes(d));
}

function walk(dir) {
  const files = fs.readdirSync(dir);

  for (const file of files) {
    const fullPath = path.join(dir, file);
    const relPath = path.relative(ROOT, fullPath);

    // Skip node_modules
    if (relPath.includes("node_modules")) continue;

    if (IGNORE_DIRS.some(dir => relPath.startsWith(dir))) {
      continue;
    }

    if (IGNORE_FILES.includes(file)) {
      continue;
    }

    const stat = fs.statSync(fullPath);

    // Check English names
    if (!isEnglishName(file)) {
      errors.push(`❌ Non-English file/folder name: ${relPath}`);
    }

    if (stat.isDirectory()) {
      walk(fullPath);
    
      if (shouldCheckIndex(relPath)) {
        const indexPath = path.join(fullPath, "index.md");
        if (!fs.existsSync(indexPath)) {
          warnings.push(`⚠️ Missing index.md in ${relPath}`);
        }
      }
    } else {
      if (!isMarkdown(file)) {
        // warnings.push(`⚠️ Non-markdown file detected: ${relPath}`);
        continue;
      } else {
        checkFileName(fullPath);

        const content = fs.readFileSync(fullPath, "utf-8");

        // Check broken relative links
        const linkRegex = /\[.*?\]\((.*?)\)/g;
        let match;

        while ((match = linkRegex.exec(content))) {
          const link = match[1];

          if (link.startsWith("./") || link.startsWith("../")) {
            const target = path.resolve(path.dirname(fullPath), link);
            if (!fs.existsSync(target)) {
              errors.push(`❌ Broken link in ${relPath} → ${link}`);
            }
          }
        }

        // Check images
        const imgRegex = /!\[.*?\]\((.*?)\)/g;
        while ((match = imgRegex.exec(content))) {
          const img = match[1];

          if (!img.startsWith("http")) {
            const imgPath = path.resolve(path.dirname(fullPath), img);
            if (!fs.existsSync(imgPath)) {
              errors.push(`❌ Missing image in ${relPath} → ${img}`);
            }
          }
        }
      }
    }
  }
}

// Protected files check (based on git diff if available)
function checkProtectedFiles() {
  try {
    const changedFiles = fs.readFileSync("changed_files.txt", "utf-8")
      .split("\n")
      .filter(Boolean);

    for (const file of changedFiles) {
      for (const protectedPath of PROTECTED_PATHS) {
        if (file.startsWith(protectedPath)) {
          errors.push(`❌ Protected file modified: ${file}`);
        }
      }
    }
  } catch {
    warnings.push("⚠️ Could not verify protected files");
  }
}

// Run
// walk(ROOT);
function getChangedFiles() {
  try {
    return fs.readFileSync("changed_files.txt", "utf-8")
      .split("\n")
      .filter(f => f.endsWith(".md"));
  } catch {
    return [];
  }
}

const changedFiles = getChangedFiles();

for (const file of changedFiles) {
  const fullPath = path.join(ROOT, file);

  if (fs.existsSync(fullPath)) {
    checkFileName(fullPath);

    const content = fs.readFileSync(fullPath, "utf-8");

    // run link + image checks here
  }
}
checkProtectedFiles();

// Output
console.log("\n## 🤖 Lint Report\n");

if (errors.length) {
  console.log("### ❌ Errors");
  errors.forEach(e => console.log(e));
}

if (warnings.length) {
  console.log("\n### ⚠️ Warnings");
  warnings.forEach(w => console.log(w));
}

if (!errors.length && !warnings.length) {
  console.log("✅ All checks passed!");
}

// Exit code
if (errors.length) {
  process.exit(1);
}

File: scripts/lint.py
import os
import re
import sys

ROOT = os.getcwd()

# -----------------------------
# Config
# -----------------------------
PROTECTED_PATHS = [
    "_config.yml",
    "_data",
    "_layouts",
    "scripts",
    "README.md",
    "CODEOWNERS",
]

ALLOWED_PROTECTED_FILES = {
    "_data/fr.yml",
}

ALLOWED_EXTENSIONS = {".md", ".jpg", ".jpeg", ".png"}

errors = []
warnings = []

# -----------------------------
# Input layer (DO NOT FILTER HERE)
# -----------------------------
def read_changed_files():
    try:
        with open("changed_files.txt", "r", encoding="utf-8") as f:
            return [line.strip() for line in f if line.strip()]
    except FileNotFoundError:
        return []


# -----------------------------
# Validation helpers
# -----------------------------
def check_file_name(file_path):
    name = os.path.basename(file_path)

    if name == "index.md":
        return

    base = os.path.splitext(name)[0]

    if not (
        re.match(r"^\d{3}-[a-z0-9-]+$", base)
        or re.match(r"^\d+\.[a-z0-9-]+$", base)
    ):
        errors.append(f"❌ Invalid file name: {file_path}")


def check_frontmatter(file_path, content):
    name = os.path.basename(file_path)

    if name == "README.md":
        return

    if not content.startswith("---"):
        errors.append(f"❌ Missing frontmatter in {file_path}")
        return

    parts = content.split("---")

    if len(parts) < 3:
        errors.append(f"❌ Invalid frontmatter format in {file_path}")
        return

    frontmatter = parts[1]

    if "layout: default" not in frontmatter:
        errors.append(f'❌ Missing "layout: default" in {file_path}')

    title_match = re.search(r"title:\s*(.+)", frontmatter)

    if not title_match:
        errors.append(f'❌ Missing "title" in {file_path}')
    elif not title_match.group(1).strip():
        errors.append(f'❌ Empty "title" in {file_path}')


def check_links_and_images(file_path, content):
    dir_path = os.path.dirname(file_path)

    # Links
    for match in re.finditer(r"\[.*?\]\((.*?)\)", content):
        link = match.group(1)

        if link.startswith("./") or link.startswith("../"):
            target = os.path.abspath(os.path.join(dir_path, link))
            if not os.path.exists(target):
                errors.append(f"❌ Broken link in {file_path} → {link}")

    # Images
    for match in re.finditer(r"!\[.*?\]\((.*?)\)", content):
        img = match.group(1)

        if not img.startswith("http"):
            img_path = os.path.abspath(os.path.join(dir_path, img))
            if not os.path.exists(img_path):
                errors.append(f"❌ Missing image in {file_path} → {img}")


def check_protected_files():
    try:
        with open("changed_files.txt", "r", encoding="utf-8") as f:
            changed = [line.strip() for line in f if line.strip()]

        for file in changed:
            # Allow explicitly permitted files
            if file in ALLOWED_PROTECTED_FILES:
                continue

            for protected in PROTECTED_PATHS:
                if file.startswith(protected):
                    errors.append(f"❌ Protected file modified: {file}")

    except FileNotFoundError:
        warnings.append("⚠️ Could not verify protected files")


def run_checks(file_path, content):
    check_file_name(file_path)
    check_frontmatter(file_path, content)
    check_links_and_images(file_path, content)


# -----------------------------
# Report
# -----------------------------
def build_report():
    lines = []
    lines.append("## 🤖 Lint Report\n")

    if errors:
        lines.append("### ❌ Errors")
        lines.extend(errors)

    if warnings:
        lines.append("\n### ⚠️ Warnings")
        lines.extend(warnings)

    if not errors and not warnings:
        lines.append("✅ All checks passed!")

    return "\n".join(lines)


def write_outputs(report):
    # Write file
    with open("report.txt", "w", encoding="utf-8") as f:
        f.write(report)

    # Export to GitHub Actions env
    github_env = os.environ.get("GITHUB_ENV")
    if github_env:
        with open(github_env, "a", encoding="utf-8") as f:
            f.write(f"LINT_REPORT<<EOF\n{report}\nEOF\n")


# -----------------------------
# Main
# -----------------------------
def main():
    changed_files = read_changed_files()

    if not changed_files:
        print("❌ No changed files detected.")
        return 1

    for file in changed_files:
        full_path = os.path.join(ROOT, file)
        ext = os.path.splitext(file)[1].lower()

        # -----------------------------
        # 1. Validate extension
        # -----------------------------
        if ext not in ALLOWED_EXTENSIONS:
            errors.append(f"❌ Invalid file type: {file}")
            continue

        # -----------------------------
        # 2. Markdown validation
        # -----------------------------
        if ext == ".md":
            if os.path.exists(full_path):
                with open(full_path, "r", encoding="utf-8") as f:
                    content = f.read()

                run_checks(full_path, content)
            else:
                errors.append(f"❌ Missing markdown file: {file}")

        # -----------------------------
        # 3. Image validation (placeholder)
        # -----------------------------
        elif ext in {".jpg", ".jpeg", ".png"}:
            if not os.path.exists(full_path):
                errors.append(f"❌ Missing image file: {file}")

    check_protected_files()

    report = build_report()

    print(report)
    write_outputs(report)

    return 1 if errors else 0


if __name__ == "__main__":
    sys.exit(main())
